src/expr/src/scalar/mod.rs

```diff
@@ -890,6 +897,114 @@ impl MirScalarExpr {
+            // LIST[[[f1, f2], [f3, f4]], [[f5, f6], [f7, f8]]] [2][n][2] --> LIST[f6, f8] [n]
+            // We iterate over the index_exprs and remove literals, but keep non-literals.
+            // When we encounter a non-literal, we need to dig into the nested ListCreates:
+            // `list_creates` will contain all the ListCreates of the current level. If an element of
+            // `list_creates` is not actually a ListCreate, then we break out of the loop.
+            // When we remove a literal, we need to partial-evaluate all ListCreates
+            // that are at the current level (except those that disappeared due to
+            // literals at earlier levels), and change each element in `list_creates`
+            // to the result of the partial evaluation.
+            let top_list_create_mv = &mut (*top_list_create);
+            let mut list_creates = vec![top_list_create_mv];
+            let mut i = 0;
+            while i < index_exprs.len() && list_creates.iter().all(|lc| is_list_create_call(lc)) {
+                if index_exprs[i].is_literal_ok() {
+                    // We can remove this index.
+                    let removed_index = index_exprs.remove(i);
+                    let index_i64 = match removed_index.as_literal().unwrap().unwrap() {
+                        Datum::Int64(sql_index_i64) => sql_index_i64 - 1,
+                        _ => unreachable!(), // always an Int64, see plan_index_list
+                    };
+                    // substitute each list_create in list_creates by its `index`th argument (or null)
+                    for list_create in &mut list_creates {
+                        let list_create_args = match list_create {
+                            MirScalarExpr::CallVariadic {
+                                func: VariadicFunc::ListCreate { .. },
+                                exprs,
+                            } => exprs,
+                            _ => unreachable!(), // func cannot be anything else than a ListCreate
+                        };
+                        // ListIndex gives null on an out-of-bounds index
+                        if index_i64 >= 0 && index_i64 < list_create_args.len().try_into().unwrap()
+                        {
+                            let index: usize = index_i64.try_into().unwrap();
+                            **list_create = list_create_args.swap_remove(index);
+                        } else {
+                            let typ = list_create_type(list_create);
+                            **list_create = MirScalarExpr::literal_null(typ);
+                        }
+                    }
+                } else {
+                    // We can't remove this index.
+                    // Increment i only in this branch, because the other removes
+                    // the current element of index_exprs.
+                    i += 1;
+                    // Change list_creates to all the arguments of all the list_creates.
```

Comment:
What is happening here is we can't modify any subexpression at this level, so we are shifting the cursors to point at the subexpressions of every expression that a cursor is currently pointing at.

Reply:
Reworded the comment